xapian-letor/ranker/listnet_ranker.cc

```diff
@@ -0,0 +1,264 @@
+
+    int list_length = feature_vectors.size();
+
+    // probability distribution, y is ground truth, while z is predict score
+    vector<double> prob_y;
+    vector<double> prob_z;
+    double expsum_y = 0.0;
+    double expsum_z = 0.0;
+
+    for (int i = 0; i < list_length; i++) {
+        expsum_y += exp(feature_vectors[i].get_label());
+        expsum_z += exp(calculateInnerProduct(new_parameters,feature_vectors[i].get_fvals()));
+    }
+
+    for (int i = 0; i < list_length; i++) {
+        prob_y.push_back(exp(feature_vectors[i].get_label())/expsum_y);
+        prob_z.push_back(exp(calculateInnerProduct(new_parameters,feature_vectors[i].get_fvals()))/expsum_z);
+
+    }
+
+    vector< vector<double> > prob;
+    prob.push_back(prob_y);
+    prob.push_back(prob_z);
+
+	return prob;
+}
+
+// Equation (3) in paper.
+static vector<double>
+calculateGradient(vector<FeatureVector> feature_vectors, vector< vector<double> > prob) {
+
+	vector<double> gradient(feature_vectors[0].get_fcount(),0);
+	int list_length = feature_vectors.size();
+
+	vector<double> prob_y(prob[0]);
+	vector<double> prob_z(prob[1]);
+
+	for (int i = 0; i < list_length; i++) {
+		vector<double> fvals = feature_vectors[i].get_fvals();
+
+		for (int k = 0; k < int(fvals.size()); k++){
+
+			double first_term = - prob_y[i] * fvals[k];
+			gradient[k] += first_term;
+
+			double second_term = prob_z[i] * fvals[k];
+			gradient[k] += second_term;
+
+		}
+	}
+
+	return gradient;
+}
+
+static void
+updateParameters(vector<double> & new_parameters, vector<double> gradient, double learning_rate) {
+	int num = new_parameters.size();
+
+	for (int i = 0; i < num; i++){
+		new_parameters[i] -= gradient[i] * learning_rate;
+	}
+
+}
+
+static void
+batchLearning(const std::vector<FeatureVector> & fvv, vector<double> & new_parameters, double learning_rate) {
+
+	// initialize Probability distributions of y and z
+	vector< vector<double> > prob = initializeProbability(fvv, new_parameters);
+	// compute gradient
+	vector<double> gradient = calculateGradient(fvv, prob);
+	// update parameters: w = w - gradient * learningRate
+	updateParameters(new_parameters, gradient, learning_rate);
+
+}
+
+void
+ListNETRanker::train_model() {
+
+	std::cout << "ListNET model begin to train..." << endl;
+
+	// get the training data
+	std::vector<Xapian::FeatureVector> fvv = get_traindata();
+	int fvv_len = fvv.size();
+
+	int feature_cnt = -1;
+	if (fvv_len != 0) {
+		feature_cnt = fvv[0].get_fcount();
+	}
+	else {
+		std::cout << "The training data is NULL!" << endl;
+		exit(1); // TODO: Throw exception
+	}
+
+	// initialize the parameters for neural network
+	vector<double> new_parameters(feature_cnt, 0.0);
+
+	// iterations
+	for (int iter_num = 1; iter_num < iterations; ++iter_num) {
+
+		batchLearning(fvv, new_parameters, learning_rate);
+	}
+
+    parameters = new_parameters;
+
+}
+
+void
+ListNETRanker::save_model_to_file(const char* output_filename) {
```

Comment:
I think file formats are actually internal developer documentation, so we don't need them in the user guide. I'd leavea them in doc comments at least for now. (They might go in the nascent developer guide at some point, I guess, along with documentation on the database backend file formats, for instance.)